libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs

```diff
@@ -417,7 +425,37 @@ private void InitializeModel()
             byte[] snapShotByteArray = Encoding.UTF8.GetBytes(content);
 
             // Create label resolver
-            _resolver = orchestrator.CreateLabelResolver(snapShotByteArray);
+            _resolver = _orchestratorDictionaryEntry.Orchestrator.CreateLabelResolver(snapShotByteArray);
```

Comment:
So in this method we check for null, and if _resolver is null, we do heavy I/O and then set the resolver. CAn you guarantee that only one instance of this class is going to have the initialize model executed simultaneously? I assume not. If not, you would have a race condition here, where in the best case scenario, multiple threads do all the processing in parallel and would all be setting the same variable. Adding lazy initialization or some kind of protection for race conditions would remove that risk. Happy to chat about lifecycle / access patterns of this class to understand what the actual risks are, since I'm just observing in vacuum without as much context.

Reply:
@carlosscastro, yes, only one instance is going to initialize the model, code is run with ConcurrentDictionary.GetOrAdd that's thread safe.

Reply:
It's possible that an OrchestratorRecognizer object is constructed using the default constructor with necessary properties set. The object's RecognizeAsync() method is then called by a pool of threads. In such scenario, the InitializeModel() function can become a race condition for creating the _resolver object, albeit the Orchestrator object creation is thread safe behind the ConcurrentDictionary container.
One solution is to move InitializeModel() call out of RecognizeAsync(), but it may break the "public" API contract.
Another solution is to put all race condition inside InitializeModel() and make it thread safe. Will implement the second option.